x/qgb/README.md

```diff
@@ -1,50 +1,320 @@
+
+| Name                   | Key                                                                                                                                                             |
+|------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| EarliestAvailableAttestationNonce | [`[EarliestAvailableAttestationNonce]`](https://github.com/celestiaorg/celestia-app/blob/9bf0cf1dd9ce31a3fecb51310c3913820b21a8c2/x/qgb/types/keys.go#L35-L37)  |
+
+The earliest attestation nonce can be set to the QGB state machine store using the [`SetEarliestAvailableAttestationNonce(...)`](https://github.com/celestiaorg/celestia-app/blob/9bf0cf1dd9ce31a3fecb51310c3913820b21a8c2/x/qgb/keeper/keeper_attestation.go#L104-L110) method, and retrieved using the [`GetEarliestAvailableAttestationNonce(...)`](https://github.com/celestiaorg/celestia-app/blob/9bf0cf1dd9ce31a3fecb51310c3913820b21a8c2/x/qgb/keeper/keeper_attestation.go#L89-L102).
+
+To check if the earliest attestation nonce is defined in store, use the [`CheckEarliestAvailableAttestationNonce(...)`](https://github.com/celestiaorg/celestia-app/blob/9bf0cf1dd9ce31a3fecb51310c3913820b21a8c2/x/qgb/keeper/keeper_attestation.go#L81-L87) method.
 
-The data commitment window is defined as a governance parameter that can be changed. Currently, it is set to 400.
+## State Transitions
 
-### Panics
+### End Block
+
+During the `EndBlock` step, we're executing the [logic](https://github.com/celestiaorg/celestia-app/blob/9bf0cf1dd9ce31a3fecb51310c3913820b21a8c2/x/qgb/abci.go#L28-L35) that handles the generating new valsets, new data commitments, and prunes when needed.
+
+### Valset handler
+
+A new valset is generated by the valset [`handler`](https://github.com/celestiaorg/celestia-app/blob/9bf0cf1dd9ce31a3fecb51310c3913820b21a8c2/x/qgb/abci.go#L84-L135) in the following cases:
+
+#### No valset in store
+
+When `EndBlock` is executed straight after genesis, the store doesn't have any valset created yet. So, it [checks](https://github.com/celestiaorg/celestia-app/blob/9bf0cf1dd9ce31a3fecb51310c3913820b21a8c2/x/qgb/abci.go#L85-L93) wether there is any valset in store, and [generates](https://github.com/celestiaorg/celestia-app/blob/9bf0cf1dd9ce31a3fecb51310c3913820b21a8c2/x/qgb/abci.go#L123-L134) a new one afterwards representing the initial validator set.
+
+#### Validator starts unbonding
+
+When a validator starts unbonding, the `LatestUnbondingBlockHeight` gets updated in a hook with the current block number. Then, inside `EndBlock`, we [check](https://github.com/celestiaorg/celestia-app/blob/9bf0cf1dd9ce31a3fecb51310c3913820b21a8c2/x/qgb/abci.go#L123) whether the `LatestUnbondingBlockHeight` corresponds to the current block height. If so, then we generate a new valset that doesn't contain that unbonding validator.
```

Comment:
We could do that, but we don't want to expect from a validator to sign if he left the validator set. Also, when we will want to implement slashing, it wouldn't make sense to slash for liveness for a validator that started  unbonding

Reply:
Even with this logic, we'd still expect situations where we're asking a validator to sign an attestation even although they are not part of the validator set. Imagine if they didn't unbond or anything but we're simply kicked out because another validator delegated enough to get itself in the set and kick out the lowest staked validator

Reply:
Got it, but let's imagine the case where one validator which has 15% of the staking power leaves the validator set, then another validator joins with 15%. Then, if we don't have the unbonding hook, no change will happen to the valset because there was no power change

Reply:
Ah measurement of the power change is net? That doesn't sound too safe/correct

Reply:
It wouldn't need to be caught because its just a redelegation. 
We're even normalizing the powers to avoid creating valsets in the case of inflation or...

Reply:
Ok so just to confirm power change isn't just the total power now compared to the total power before? It actually looks to see what validators had what power?

Reply:
So is the unbonding hook still necessary?

Reply:
yes, because if that sum is not significant, no change to the valset will happen, even if the validator unbonded. Meaning we will still keep waiting for them to sign or think we can slash them